okio/src/main/kotlin/okio/Utf8.kt

```diff
@@ -110,89 +110,354 @@ fun String.utf8Size(beginIndex: Int = 0, endIndex: Int = length): Long {
     }
-    b0 and 0xf0 == 0xe0 -> {
-      // 0x1110xxxx
-      codePoint = b0 and 0x0f
-      byteCount = 3 // 16 bits (4 + 6 + 6).
-      min = 0x800
+  }
+
+  val b0 = this[beginIndex]
+  val b1 = this[beginIndex + 1]
+  if (!isUtf8Continuation(b1)) {
+    yield(REPLACEMENT_CODE_POINT)
+    return 1
+  }
+  val b2 = this[beginIndex + 2]
+  if (!isUtf8Continuation(b2)) {
+    yield(REPLACEMENT_CODE_POINT)
+    return 2
+  }
+
+  val codePoint =
+    (mask3
+        xor (b2.toInt())
+        xor (b1.toInt() shl 6)
+        xor (b0.toInt() shl 12))
+
+  when {
+    codePoint < 0x800 -> {
+      yield(REPLACEMENT_CODE_POINT) // Reject overlong code points.
     }
-    b0 and 0xf8 == 0xf0 -> {
-      // 0x11110xxx
-      codePoint = b0 and 0x07
-      byteCount = 4 // 21 bits (3 + 6 + 6 + 6).
-      min = 0x10000
+    codePoint in 0xd800..0xdfff -> {
+      yield(REPLACEMENT_CODE_POINT) // Reject partial surrogates.
     }
     else -> {
-      // We expected the first byte of a code point but got something else.
-      return REPLACEMENT_CHARACTER
+      yield(codePoint)
     }
   }
+  return 3
+}
 
-  if (size < index + byteCount) {
-    // Not enough remaining data to interpret as a code point.
-    return REPLACEMENT_CHARACTER
-  }
+/* ktlint-disable indent */
+internal val mask4 =
+    ((0xf0.toByte() shl 18) xor
+     (0x80.toByte() shl 12) xor
+     (0x80.toByte() shl 6) xor
+     (0x80.toByte().toInt()))
+/* ktlint-enable indent */
 
-  // Read the continuation bytes. If we encounter a non-continuation byte, the sequence thus far is
-  // decoded as the replacement character.
-  for (i in 1 until byteCount) {
-    val b = this[index + i]
-    if (b and 0xc0 == 0x80) {
-      // 0x10xxxxxx
-      codePoint = codePoint shl 6
-      codePoint = codePoint or (b and 0x3f)
+internal inline fun ByteArray.process4Utf8Bytes(
+  beginIndex: Int,
+  endIndex: Int,
+  yield: (Int) -> Unit
+): Int {
+  if (endIndex <= beginIndex + 3) {
+    // At least 3 bytes remaining
+    yield(REPLACEMENT_CODE_POINT)
+    if (endIndex <= beginIndex + 1 || !isUtf8Continuation(this[beginIndex + 1])) {
+      // Only 1 byte remaining - underflow
+      // Or 2nd byte is not a continuation - malformed
+      return 1
+    } else if (endIndex <= beginIndex + 2 || !isUtf8Continuation(this[beginIndex + 2])) {
+      // Only 2 bytes remaining - underflow
+      // Or 3rd byte is not a continuation - malformed
+      return 2
     } else {
-      return REPLACEMENT_CHARACTER
+      // Only 3 bytes remaining - underflow
+      return 3
     }
   }
 
-  return when {
+  val b0 = this[beginIndex]
+  val b1 = this[beginIndex + 1]
+  if (!isUtf8Continuation(b1)) {
+    yield(REPLACEMENT_CODE_POINT)
+    return 1
+  }
+  val b2 = this[beginIndex + 2]
+  if (!isUtf8Continuation(b2)) {
+    yield(REPLACEMENT_CODE_POINT)
+    return 2
+  }
+  val b3 = this[beginIndex + 3]
+  if (!isUtf8Continuation(b3)) {
+    yield(REPLACEMENT_CODE_POINT)
+    return 3
+  }
+
+  val codePoint =
```

Comment:
I’m really lost on how this works . . . the codepoint shouldn’t have any masks?

Reply:
I'll try and come up with a good way of documenting these calculations. It's not as straight forward as it was before.

Reply:
I've added a lengthy - probably too much so - section on decoding of UTF-8 bytes into a code point. Hopefully it clears up the mask + xor calculations used in these process functions.